prog/vnet/update_load_balancer_node.rb

```diff
@@ -47,7 +47,8 @@ def generate_lb_based_nat_rules
     end
 
     ipv4_prerouting = if load_balancer.ipv4_enabled?
-      load_balancer.active_vm_ports.uniq { |row| row.load_balancer_port.id }.map do |vm_port|
+      load_balancer.active_vm_ports.uniq { |row| row.load_balancer_port.id }
+        .sort_by { |vm_port| vm_port.load_balancer_port.src_port }.map do |vm_port|
```

Comment:
This looks like it has an N+1 query issue, even before this change, not just here but in the rest of the method.

Near the top of this method, you should add:

```ruby
load_balancer.active_vm_ports{|ds| ds.eager(:load_balancer_port, load_balancer_vm: :vm)}
```

That will make sure that `active_vm_ports` has eagerly loaded the related associations.

In general, for performance reasons, you should prefer direct column access over indirect column access through an associated object.  For example, `load_balancer_port.id` should be changed to `load_balancer_port_id`.  Only go through the association if the data you need is not in the current object (e.g. `src_port` access requires going through the association).

```suggestion
      load_balancer
        .active_vm_ports
        .uniq(&:load_balancer_port_id)
        .sort_by { |vm_port| vm_port.load_balancer_port.src_port }
        .map do |vm_port|
```

similar change below.

Reply:
thanks for the insight I fixed it. I just added the eager for load_balancer_port and removed the other because I was not using the load_balancer_vm or its vm.